share/lib/python/neuron/hclass3.py

```diff
@@ -4,38 +4,116 @@
 # h.anyclass methods may be overridden. If so the base method can be called
 # using the idiom self.basemethod = self.baseattr('methodname')
 # ------------------------------------------------------------------------------
-      
-from neuron import h, hoc
+
+__all__ = ["hclass", "nonlocal_hclass", "HocBaseObject"]
+
+from . import h, hoc
 import nrn
+import sys
+
+def assert_not_hoc_composite(cls):
+    """
+    Asserts that a class is not directly composed of multiple HOC types.
+    """
+    hoc_bases = [b for b in cls.__bases__ if issubclass(b, hoc.HocObject)]
+    if len(hoc_bases) > 1:
+        bases = ", ".join(b.__name__)
+        cname = cls.__name__
+        raise TypeError(f"Composition of {bases} HocObjects not allowed in {cname}")
+
+def hclass(hoc_type):
+    """
+    Class factory for subclassing HOC types.
+
+    Example:
+
+    ..code-block:: python
 
-#avoid syntax error if compiled by python 2
-exec('''
-class MetaHocObject(type):
-  """Provides Exception for Inheritance of multiple HocObject"""
-  def __new__(cls, name, bases, attrs):
-    #print cls, name, bases
-    m = []
-    for b in bases:
-      if issubclass(b, hoc.HocObject):
-        m.append(b.__name__)
-    if (len(m) > 1):
-      raise TypeError('Multiple Inheritance of HocObject in %s' % name
-        + ' through %s not allowed' % ','.join(m))
-      #note that join(b.__name__ for b in m) is not valid for Python 2.3
-  
-    return type.__new__(cls, name, bases, attrs)
-
-def hclass(c):
-    """Class factory for subclassing h.anyclass. E.g. class MyList(hclass(h.List)):..."""
-    if c == h.Section :
+        import neuron
+
+        class MyVector(neuron.hclass(neuron.h.Vector)):
+            pass
+
+    :param hoc_type: HOC types/classes such as ``h.List``, ``h.NetCon``, ``h.Vector``, ...
+    :type hoc_type: :class:`hoc.HocObject`
+    :deprecated: Inherit from :class:`~neuron.HocBaseObject` instead.
+    """
+    if hoc_type == h.Section:
         return nrn.Section
-    class hc(hoc.HocObject, metaclass=MetaHocObject):
-    #class hc(hoc.HocObject):
-        def __new__(cls, *args, **kwds):
-            kwds2 = {'hocbase': cls.htype}
-            if 'sec' in kwds:
-                kwds2['sec'] = kwds['sec']
-            return hoc.HocObject.__new__(cls, *args, **kwds2)
-    setattr(hc, 'htype', c)
+    try:
+        hc = type("hc", (HocBaseObject,), {}, hoc_type=hoc_type)
+    except TypeError:
+        raise TypeError("Argument is not a valid HOC type.") from None
     return hc
-''')
+
+def nonlocal_hclass(hoc_type, module_name, name=None):
```

Comment:
Are there any cases where we'd want to use `hclass` instead of this? I don't fully understand the relationship, but if this is better, shouldn't it be the default behavior?

Reply:
I didn't want to break backwards compatibility in my proposal as I don't know which role `hclass` plays and couldn't estimate an impact. If you don't mind that the factory class' `__name__`, `__module__` and `__qualname__` change then yes, this should just be a general improvement that helps other modules like `inspect`, `pickle` - really just anything that tries to import these classes based on strings.

Reply:
Curiously, there are two models ([206400](https://modeldb.yale.edu/206400) and [244412](https://modeldb.yale.edu/244412)) in ModelDB that import hclass, but neither uses it. The only place it is used in the NEURON source appears to be in the `_subclass` test.

Reply:
As far as the signature goes: I would say it would be okay if there was a way to provide a reasonable default for `module_name` so that you could still call it exactly as before and get something useful.

Reply:
Ok, I notice that in my PR I did mix 2 approaches I tried (and `nonlocal_hclass` will currently error):

* Passing the module object and toplevel variable name: `nonlocal_hclass(h.Vector, globals(), "MyVector")`
  * Ensures that the module level variable name and class `__name__` match
  * Mutates the given module (well, its `__dict__`)
  * Not that legible since there's no explicit declaration of `MyVector` in the module.
* Passing just the module name `MyVector = nonlocal_hclass(h.Vector, __name__, "MyVector")`
  * Is more clear
  * Does allow a mismatch in names in the case of `MyVeectr = nonlocal_hclass(h.Vector, __name__, "MyVector")`
  
 Personally I'd go for option 2 for clarity, the `__name__` of the produced class is not THAT important to start doing funny things for.

Reply:
Why wouldn't you want the name to match the name of the underlying NEURON entity? (i.e. why not default to `Vector` here, if not always making it `Vector`)

Reply:
Well if you have `MyVector = hclass(h.Vector)` then the `__name__` of the produced class should be `MyVector`, not `Vector`, as the class would be importable as `that_module.MyVector`, not `that_module.Vector`. There's no use case for making mistakes against that so at one point I experimented with approach 1 that would prevent mistakes and duplicate information (you enter the name twice, once as `MyVector = ...` and then later as an arg to that function call)

I'm OK with either approaches, slightly tilted in favor of the one that doesn't use `globals()`.

Reply:
But ok, I guess your earlier reply that I missed settles it:

> As far as the signature goes: I would say it would be okay if there was a way to provide a reasonable default for module_name so that you could still call it exactly as before and get something useful.

I'll make one factory: `hclass(hoc_type, module_name=None, name=None)` and use default  `module_name = neuron.hclass3` and `name = hoc_type.hname()[:2]`.

Reply:
Fixed in 3722763